src/ucp/rma/basic_rma.c

```diff
@@ -26,142 +30,242 @@
-        /* Should be replaced with bcopy */
+    if (req->send.length <= ucp_ep_config(ep)->bcopy_thresh) {
         packed_len = ucs_min(req->send.length, rma_config->max_put_short);
-        status = uct_ep_put_short(ep->uct_eps[req->send.lane],
+        status = UCS_PROFILE_CALL(uct_ep_put_short,
+                                  ep->uct_eps[req->send.lane],
                                   req->send.buffer,
                                   packed_len,
                                   req->send.rma.remote_addr,
                                   uct_rkey);
     } else {
-        /* We don't do it right now, but in future we have to add
-         * an option to use zcopy
-         */
         ucp_memcpy_pack_context_t pack_ctx;
         pack_ctx.src    = req->send.buffer;
         pack_ctx.length = ucs_min(req->send.length, rma_config->max_put_bcopy);
-        packed_len = uct_ep_put_bcopy(ep->uct_eps[req->send.lane],
+        packed_len = UCS_PROFILE_CALL(uct_ep_put_bcopy,
+                                      ep->uct_eps[req->send.lane],
                                       ucp_memcpy_pack,
                                       &pack_ctx,
                                       req->send.rma.remote_addr,
                                       uct_rkey);
         status = (packed_len > 0) ? UCS_OK : (ucs_status_t)packed_len;
     }
 
-    if ((status == UCS_OK) || (status == UCS_INPROGRESS)) {
-        req->send.length -= packed_len;
-        if (req->send.length == 0) {
-            ucp_request_put(req, UCS_OK);
-            return UCS_OK;
-        }
+    return ucp_rma_request_advance(req, packed_len, status);    
+}
 
-        req->send.buffer += packed_len;
-        req->send.rma.remote_addr += packed_len;
-        return UCS_INPROGRESS;
-    } else {
+static ucs_status_t 
+ucp_progress_get_inner(ucp_request_t *req, uct_rkey_t uct_rkey, 
+                       ucp_ep_rma_config_t *rma_config)
+{
+    ucp_ep_t *ep = req->send.ep;
+    ucs_status_t status;
+    size_t frag_length;
+
+    ++req->send.uct_comp.count;
+    frag_length = ucs_min(rma_config->max_get_bcopy, req->send.length);
+    status = UCS_PROFILE_CALL(uct_ep_get_bcopy,
+                              ep->uct_eps[req->send.lane],
+                              (uct_unpack_callback_t)memcpy,
+                              (void*)req->send.buffer,
+                              frag_length,
+                              req->send.rma.remote_addr,
+                              uct_rkey,
+                              &req->send.uct_comp);
+    if (status <= 0) {
+        --req->send.uct_comp.count;
+    }
+
+    return ucp_rma_request_advance(req, frag_length, status);    
+}
+
+static ucs_status_t ucp_progress_get_nbi(uct_pending_req_t *self)
+{
+    uct_rkey_t uct_rkey;
+    ucp_ep_rma_config_t *rma_config;
+    ucs_status_t status;
+
+    ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct);
```

Comment:
pls put the local variables which are initialized with values in the very beginning of the function